### PR TITLE
Rust demo

### PR DESCRIPTION
**Pull request by szg251**
- Implement Rust demo
- Add darwin to Hercules CI

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -13,10 +13,15 @@
...
 
     lbf.url = "github:mlabs-haskell/lambda-buffers";
+    flake-lang.url = "github:mlabs-haskell/flake-lang.nix";
     haskell-nix.follows = "lbf/haskell-nix";
     iohk-nix.follows = "lbf/iohk-nix";
     ctl.follows = "lbf/ctl";
     plutarch.follows = "lbf/plutarch";
+    crane.url = "github:ipetkov/crane";
+
+    plutip.url = "github:mlabs-haskell/plutip/1bf0b547cd3689c727586abb8385c008fb2a3d1c";
```
Comment (szg251) [inline flake.nix:23]: I tried using pkgsForCtl, but unfortunately that failed on Mac (even on my local machine).
This is the exact version that CTL uses (unfortunately I couldn't use `follows = ctl/plutip` for some misterious reasons)

File: flake.nix [hunk 2]
```diff
@@ -30,9 +35,11 @@
         ./validation/demo-plutarch/build.nix
         ./validation/demo-plutustx/build.nix
         ./transactions/demo-ctl/build.nix
+        ./transactions/demo-rust/build.nix
       ];
       debug = true;
-      systems = [ "x86_64-linux" "x86_64-darwin" ];
+      systems = [ "x86_64-linux" ]; # TODO(szg251): fix darwin CI build and add "x86_64-darwin"
```
Comment (szg251) [inline flake.nix:41]: Funnily (yeah, it wasn't that funny) I could build and run tests on a local x86_64-darwin machine, but it fails on CI.
It looks like there's an upstream issue with plutip and ogmios

File: hercules-ci.nix [hunk 3]
```diff
@@ -17,5 +17,5 @@
     };
   };
 
-  herculesCI.ciSystems = [ "x86_64-linux" ];
+  herculesCI.ciSystems = [ "x86_64-linux" ]; # TODO(szg251): fix darwin CI build and add "x86_64-darwin"
```
Comment (szg251) [inline hercules-ci.nix:20]: I wanted to enable darwin, because it works on my machine.
Funnily (yeah, it wasn't that funny) I could build and run tests on a local x86_64-darwin machine, but it fails on CI.
It looks like there's an upstream issue with plutip and ogmios

File: transactions/demo-rust/.config/nextest.toml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+[profile.default]
```
Comment (szg251) [inline transactions/demo-rust/.config/nextest.toml:1]: As we're using ogmios and plutip which are using some set ports/folders, parallel testing fails.
I have 3 places to configure this, so things work in every environment:
- `serial_test` crate and `#[serial]` macro for `cargo test`
- this configuration for `nextest` locally, without using nix
- `cargoNextestExtraArgs` in the `build.nix` file for nix (the above configuration is not included, `cleanCargoSource` filters the `.config` folder)

File: transactions/demo-rust/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+use lbf_demo_plutus_api::demo::plutus::{EqDatum, EqRedeemer};
+use plutus_ledger_api::plutus_data::IsPlutusData;
+use std::collections::BTreeMap;
+
+pub mod utils;
+
+const COINS_PER_UTXO_WORD: u64 = 34_482;
+
+/// Transaction that stores a EqDatum value at the Eq Validator.
+pub fn mk_lock_tx(
```
Comment (szg251) [inline transactions/demo-rust/src/lib.rs:20]: The idea here is to have a pure builder part (mk_lock_tx) and an effectful part (lock_tx_build_and_submit).
The latter one takes a Wallet (capable of signing and submitting) and Ogmios (this could be similary abstracted into an Indexer).
These are currently pretty arbitrary, more like an idea for future explorations. It's likely, that some functionality of Ogmios would be taken over by the Wallet, to be in line with CIP-30 https://cips.cardano.org/cip/CIP-30/
Also, we shouldn't see raw CSL data types at this level.

File: transactions/demo-rust/.config/nextest.toml (hunk 4 above)
Comment (bladyjoker) [reply transactions/demo-rust/.config/nextest.toml:1]: I have no idea what this is, and will probably not know what this is next time I come here. If it's needed then ok, otherwise remove everything that's not necessary

File: transactions/demo-rust/src/lib.rs
Comment (bladyjoker) [inline transactions/demo-rust/src/lib.rs:1]: I would put this in `transactions` like https://github.com/mlabs-haskell/lambda-buffers-for-cardano/blob/main/transactions/demo-ctl/src/Demo/Transactions.purs

That's where the pure transactions are.

File: transactions/demo-rust/src/lib.rs (hunk 5 above)
Comment (bladyjoker) [reply transactions/demo-rust/src/lib.rs:22]: Please use the naming we use in CTL part. If you think CTL naming is bad, which i do think it is :D, change it there and align. This repo should serve as a demonstration of how equivalent things are implemented in different ecosystems.

File: transactions/demo-rust/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            csl::tx_builder::CoinSelectionStrategyCIP2::RandomImproveMultiAsset,
+        )
+        .unwrap();
+    tx_builder.add_required_signer(own_pkh);
+
+    tx_builder
+}
+
+/// Builds and submits `mk_lock_tx`
+pub async fn lock_tx_build_and_submit(
```
Comment (bladyjoker) [inline transactions/demo-rust/src/lib.rs:126]: I love that we have the pure part, and the IO part.

I want to refactor this into a separate module though, not sure where. Why?

Here we make a bunch of choices that are arbitrary, for example, Ogmios, that are very specific to offchain orchestration that can be quite varied.

I placed that under test, as I see it as a 'test application'. Maybe under `tests/transactions.rs`?


So like `src/transactions.rs` is pure transactions that is the core of the protocol api, and `tests/transactions.rs` is a particular orchestration?

File: transactions/demo-rust/build.nix [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            ]) ++
+            (pkgs.lib.optionals pkgs.stdenv.isDarwin
+              [
+                pkgs.gcc
+                pkgs.darwin.apple_sdk.frameworks.Security
+                pkgs.darwin.apple_sdk.frameworks.SystemConfiguration
+              ]);
+          buildInputs = [
+            pkgs.openssl.dev
+          ];
```
Comment (bladyjoker) [inline transactions/demo-rust/build.nix:22]: Wait, why is this here? Shouldn't this be upstreamed? I don't want to know about this if i want to have my rust project build with darwin. Or do I?

File: flake.nix (hunk 1 above)
Comment (bladyjoker) [reply flake.nix:23]: What's the reason? And also, why do you care about using CTL's versions? You can just point at the latest of both? No need to pin for your Rust parts?

File: transactions/demo-rust/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+edition = "2021"
+
+[dependencies]
+lbf-prelude = { path = ".extras/lbf-prelude" }
+lbf-demo-config-api = { path = ".extras/lbf-demo-config-api" }
+lbf-demo-plutus-api = { path = ".extras/lbf-demo-plutus-api" }
+cardano-serialization-lib = "11.5.0"
+plutus-ledger-api = { git = "https://github.com/mlabs-haskell/plutus-ledger-api-rust", features = [
+  "lbf",
+], rev = "d66d39c949f59e543c91ee36377c93422d8e9d75" }
```
Comment (bladyjoker) [inline transactions/demo-rust/Cargo.toml:13]: Any chance we can use `version` tag? easier to upgrade i'd say

**Review by bladyjoker — COMMENTED** [review]
Woot woot, left some comments! This is so exciting

File: flake.nix (hunk 1 above)
Comment (szg251) [reply flake.nix:23]: simply because the latest commit also failed...
the whole ecosystem is broken

Comment (szg251) [reply flake.nix:23]: follows = ctl/plutip failed on ci with evaluation failure. I don't really understand, because it worked locally.

File: transactions/demo-rust/src/lib.rs (hunk 5 above)
Comment (szg251) [reply transactions/demo-rust/src/lib.rs:22]: I'll refactor the ctl side a little bit then

File: transactions/demo-rust/Cargo.toml (hunk 8 above)
Comment (szg251) [reply transactions/demo-rust/Cargo.toml:13]: That would be a good idea, now that it's more stable.
We could even start thinking about releasing this to crates.io

**Review by bladyjoker — APPROVED** [review]
